third_party/move/move-compiler-v2/src/function_checker.rs

```diff
@@ -253,6 +253,38 @@ pub fn check_access_and_use(env: &mut GlobalEnv, before_inlining: bool) {
                                 Visibility::Friend => {
                                     if callee_func.module_env.has_friend(&caller_module_id) {
                                         true
+                                    } else if callee_func.has_package_visibility() {
+                                        if callee_func.module_env.self_address()
+                                            == caller_func.module_env.self_address()
+                                        {
+                                            if callee_func
+                                                .module_env
+                                                .in_same_package(&caller_func.module_env)
+                                            {
+                                                // we should've inferred the friend declaration
+                                                panic!(
+                                                    "{} should have friend {}",
+                                                    callee_func.module_env.get_full_name_str(),
+                                                    caller_func.module_env.get_full_name_str()
+                                                );
+                                            } else {
+                                                call_package_fun_from_diff_package_error(
+                                                    env,
+                                                    sites,
+                                                    &caller_func,
+                                                    &callee_func,
+                                                );
+                                                false
+                                            }
+                                        } else {
+                                            call_package_fun_from_diff_addr_error(
```

Comment:
A package doesn't necessarily need to have all modules at the same address, so perhaps we should relax this. Though I'm not sure whether friend has a restriction like this. In general I'd think we do not need this check.

Reply:
> Friend modules must be within the same account address. (Note: this is not a technical requirement but rather a policy decision which may be relaxed later.)

From the Aptos Move book. Since we are converting to friends, we need to check the same restriction.